perma_web/perma/models.py

```diff
@@ -1273,85 +1269,208 @@ class Folder(MPTTModel):
-                self.get_descendants(include_self=True).update(owned_by=None, organization=self.parent.organization_id, sponsored_by=None)
-                if links:
-                    links.update(organization_id=self.parent.organization_id)
-            elif self.parent.sponsored_by_id:
-                self.get_descendants(include_self=True).update(owned_by=self.parent.owned_by_id, organization=None, sponsored_by_id=self.parent.sponsored_by_id)
-                if links:
-                    links.update(organization_id=None)
+                    self.tree_root_id = self.id
+                if self.created_by_id and not self.owned_by_id and not self.organization_id:
+                    self.owned_by_id = self.created_by_id
+
+                # Save and refresh from db before continuing, because we need an ID
+                super().save(*args, **kwargs)
+                self.refresh_from_db()
+                descendant_ids = [self.id]
+                descendants = Folder.objects.with_tree_fields().filter(id__in=descendant_ids)
+
             else:
-                self.get_descendants(include_self=True).update(owned_by=self.parent.owned_by_id, organization=None, sponsored_by=None)
-                if links:
-                    links.update(organization_id=None)
-            # credit users for any bonus links they are due
-            if self.parent.organization_id or self.parent.sponsored_by_id:
-                if bonus_links:
-                    user = bonus_links[0].created_by
-                    count = bonus_links.update(bonus_link=False)
-                    user.bonus_links = user.bonus_links + count
-                    user.save(update_fields=['bonus_links'])
-
-        if new or parent_has_changed:
-            # update cached paths
-            for descendant in self.get_descendants(include_self=True):
-                descendant.cached_path = descendant.get_path()
-                descendant.save()
-
-
-    class MPTTMeta:
-        order_insertion_by = ['name', 'id']
+
+                # We find descendants on the fly by inspecting parent_id.
+                # Retrieve descendant IDs before saving, while self.parent_id is still unchanged in the DB...
+                # otherwise you won't find anything, not even self!
+                if self.cached_has_children:
+                    descendant_ids = list(
+                        self.get_descendants(
+                            include_self=True
+                        ).tree_filter(
+                            tree_root_id=self.tree_root_id
+                        ).values_list(
+                            'id', flat=True
+                        )
+                    )
+                else:
+                    descendant_ids = [self.id]
+                descendants = Folder.objects.with_tree_fields().filter(id__in=descendant_ids)
+                super(Folder, self).save(*args, **kwargs)
+
+            if parent_has_changed:
```

Comment:
I think this section is probably worth documenting a tiny bit more for clarity.
Specifically, the succession of:
```python
if parent_has_changed: 
  ... 
if new or parent_has_changed:
  .... 
```

Reply:
Yes! Will do in a followup PR, which will probably also include some refactoring.

Reply:
@matteocargnelutti FWIW I had a go at making this [easier to read and reason about](https://github.com/harvard-lil/perma/blob/946ec3176154d54c2854186057ee0194776737d8/perma_web/perma/models.py#L1315-L1471) in https://github.com/harvard-lil/perma/pull/3529.